pkg/controller/pxc/controller.go

```diff
@@ -534,6 +476,113 @@ func (r *ReconcilePerconaXtraDBCluster) Reconcile(request reconcile.Request) (re
 	return rr, nil
 }
 
+func (r *ReconcilePerconaXtraDBCluster) reconcileHAProxy(cluster *api.PerconaXtraDBCluster, owner metav1.OwnerReference) error {
+	if !cluster.HAProxyEnabled() {
+		if err := r.disableHAProxyReplicasService(cluster); err != nil {
+			return errors.Wrap(err, "delete HAProxy replica service")
+		}
+
+		if err := r.disableHAProxyService(cluster); err != nil {
+			return errors.Wrap(err, "delete HAProxy service")
+		}
+
+		if err := r.disableHAProxy(cluster); err != nil {
+			return errors.Wrap(err, "delete HAProxy stateful set")
+		}
+
+		return nil
+	}
+
+	if err := r.enableHAProxy(cluster); err != nil {
+		return errors.Wrap(err, "HAProxy upgrade error")
+	}
+
+	if err := r.enableHAProxyService(cluster, owner); err != nil {
+		return errors.Wrap(err, "failed to create or update haproxy service")
+	}
+
+	if cluster.HAProxyReplicasServiceEnabled() {
+		if err := r.enableHAProxyReplicasService(cluster, owner); err != nil {
+			return errors.Wrap(err, "failed to create or update haproxy replicas")
+		}
+	} else {
+		if err := r.disableHAProxyReplicasService(cluster); err != nil {
+			return errors.Wrap(err, "delete HAProxy replica service")
+		}
+	}
+
+	return nil
+}
+
+func (r *ReconcilePerconaXtraDBCluster) enableHAProxy(cluster *api.PerconaXtraDBCluster) error {
+	sfs := statefulset.NewHAProxy(cluster)
+	ps := &cluster.Spec.HAProxy.PodSpec
+
+	return r.updatePod(sfs, ps, cluster, nil)
+}
+
+func (r *ReconcilePerconaXtraDBCluster) enableHAProxyService(cluster *api.PerconaXtraDBCluster, owner metav1.OwnerReference) error {
+	haProxyService := pxc.NewServiceHAProxy(cluster, owner)
+	haProxyService.Spec.Type = corev1.ServiceTypeClusterIP
+
+	if len(cluster.Spec.HAProxy.ServiceType) > 0 {
+		haProxyService.Spec.Type = cluster.Spec.HAProxy.ServiceType
+	}
+
+	if haProxyService.Spec.Type == corev1.ServiceTypeLoadBalancer ||
+		haProxyService.Spec.Type == corev1.ServiceTypeNodePort {
+		haProxyService.Spec.ExternalTrafficPolicy = corev1.ServiceExternalTrafficPolicyTypeCluster
+
+		if len(cluster.Spec.HAProxy.ExternalTrafficPolicy) > 0 {
+			haProxyService.Spec.ExternalTrafficPolicy = cluster.Spec.HAProxy.ExternalTrafficPolicy
+		}
+	}
+
+	return r.createOrUpdate(haProxyService)
+}
+
+func (r *ReconcilePerconaXtraDBCluster) enableHAProxyReplicasService(cluster *api.PerconaXtraDBCluster, owner metav1.OwnerReference) error {
+	haProxyServiceReplicas := pxc.NewServiceHAProxyReplicas(cluster, owner)
+
+	replicaPorts := []corev1.ServicePort{
+		{
+			Port:       3306,
+			TargetPort: intstr.FromInt(3307),
+			Name:       "mysql-replicas",
+		},
+	}
+
+	haProxyServiceReplicas.Spec.Ports = replicaPorts
```

Comment:
we can assign the array directly to this variable, instead of initializing a new array